// hullabaloo-config-manager/build/ResolvedConfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const codegen_1 = require("./codegen");
const reduceChains_1 = require("./reduceChains");
const Verifier_1 = require("./Verifier");
class ResolvedConfig {
    constructor(chains, cache) {
        this.cache = cache;
        this.babelrcDir = chains.babelrcDir;
        const reduced = reduceChains_1.default(chains, cache);
        this.dependencies = reduced.dependencies;
        this.envNames = reduced.envNames;
        this.fixedSourceHashes = reduced.fixedSourceHashes;
        this.sources = reduced.sources;
        this.unflattenedDefaultOptions = reduced.unflattenedDefaultOptions;
        this.unflattenedEnvOptions = reduced.unflattenedEnvOptions;
    }
    createVerifier() {
        return Verifier_1.default.hashAndCreate(this.babelrcDir, this.envNames, this.dependencies, this.sources, this.fixedSourceHashes, this.cache);
    }
    generateModule() {
        return codegen_1.default(this);
    }
}
exports.default = ResolvedConfig;
